bw2io/ecoinvent.py

```diff
@@ -0,0 +1,260 @@
+import re
+import zipfile
+from collections import defaultdict
+from pathlib import Path
+
+import bw2data as bd
+import ecoinvent_interface as ei
+import openpyxl
+from ecoinvent_interface.core import SYSTEM_MODELS
+from ecoinvent_interface.string_distance import damerau_levenshtein
+
+from .extractors import ExcelExtractor
+from .importers import (
+    EcoinventLCIAImporter,
+    Ecospold2BiosphereImporter,
+    SingleOutputEcospold2Importer,
+)
+
+
+def get_excel_sheet_names(file_path: Path) -> list[str]:
+    """Read XML metadata file instead of using openpyxl, which loads the whole workbook.
+
+    From https://stackoverflow.com/questions/12250024/how-to-obtain-sheet-names-from-xls-files-without-loading-the-whole-file.
+    """
+    sheets = []
+    with zipfile.ZipFile(file_path, "r") as zip_ref:
+        xml = zip_ref.read("xl/workbook.xml").decode("utf-8")
+        for s_tag in re.findall("<sheet [^>]*", xml):
+            sheets.append(re.search('name="[^"]*', s_tag).group(0)[6:])
+    return sheets
+
+
+def header_dict(array: list) -> list[dict]:
+    return [
+        {header.lower(): value for header, value in zip(array[0], row)}
+        for row in array[1:]
+    ]
+
+
+def drop_unspecified(a: str, b: str, c: str) -> tuple:
+    if c.lower() == "unspecified":
+        return (a, b)
+    else:
+        return (a, b, c)
+
+
+def pick_a_unit_label_already(obj: dict) -> str:
+    candidates = ("indicator unit", "unit", "unitname", "impact score unit")
+    for candidate in candidates:
+        if candidate in obj:
+            return candidate
+    raise KeyError("Can't find suitable column label for LCIA units")
+
+
+def import_ecoinvent_release(
+    version: str,
+    system_model: str,
+    username: str | None = None,
+    password: str | None = None,
+    lci: bool = True,
+    lcia: bool = False,
+    biosphere_name: str | None = None,
+) -> None:
+    """Import an ecoinvent LCI and optionally LCIA database.
+
+    Uses [ecoinvent_interface](https://github.com/brightway-lca/ecoinvent_interface). Auth credentials are optional as they can be set externally (see the `ecoinvent_interface` documentation), and such permanent storage is highly recommended.
+
+    The biosphere database must not exist. It is too much work to do selective updates.
+
+    Parameters
+    ----------
+    version
+        The ecoinvent release version as a string, e.g. '3.9.1'
+    system_model
+        The system model as a string in short or long form, e.g. 'apos' or 'Allocation cut-off by classification'
+    username
+        ecoinvent username
+    password
+        ecoinvent password
+    lci
+        Flag on whether to import the inventory database
+    lcia
+        Flag on whether to import the LCIA impact categories
+    biosphere_name
+        Name of database to store biosphere flows. They will be stored in the main LCI database if not specified.
+
```

Comment:
suggestion [documentation]: add examples on usage.

We should add an "Example" on how import one old and the "latest" available (3.9.1) for different system models, commenting on how to use the different flags also. I'm thinking about the pandas examples here.

Reply:
Good idea. Encodings definitely need testing, as ecoinvent releases have used `latin-1` or mixed encodings in the past, at least in some data products.